--- patch_realtek_org.c	2008-03-22 15:41:54.000000000 -0400
+++ patch_realtek.c	2008-03-22 15:48:13.000000000 -0400
@@ -144,6 +144,7 @@ enum {
	ALC882_TARGA,
	ALC882_ASUS_A7J,
	ALC885_MACPRO,
+	ALC885_APPLETV,
	ALC885_MBP3,
	ALC882_AUTO,
	ALC882_MODEL_LAST,
@@ -5050,6 +5051,15 @@ static struct snd_kcontrol_new alc882_ma
	{ } /* end */
 };

+/* AppleTV Mixer */
+static struct snd_kcontrol_new alc885_appletv_mixer[] = {
+	HDA_CODEC_VOLUME("Master Playback Volume", 0x0c, 0x0 , HDA_OUTPUT),
+	HDA_BIND_MUTE("Master Playback Switch", 0x0c, 2 , HDA_INPUT),
+	HDA_CODEC_VOLUME("Line Playback Volume", 0x0b, 7, HDA_INPUT),
+	HDA_CODEC_MUTE("Line Playback Switch", 0x0b, 7, HDA_INPUT),
+	{ } /* end */
+};
+
static struct hda_verb alc882_macpro_init_verbs[] = {
	/* Front mixer: unmute input/output amp left and right (volume = 0) */
	{0x0c, AC_VERB_SET_AMP_GAIN_MUTE, AMP_OUT_ZERO},
@@ -5161,6 +5171,15 @@ static struct hda_verb alc885_mbp3_init_
	{ }
};
 
+static struct hda_verb alc885_appletv_init_verbs[] = {
+	{0x14, AC_VERB_SET_PIN_WIDGET_CONTROL, PIN_OUT},
+	{0x14, AC_VERB_SET_AMP_GAIN_MUTE, AMP_OUT_UNMUTE},
+	{0x14, AC_VERB_SET_CONNECT_SEL, 0x00},
+	{ }
+};
+
+
+
static struct hda_verb alc882_targa_verbs[] = {
	{0x0c, AC_VERB_SET_AMP_GAIN_MUTE, AMP_IN_UNMUTE(0)},
	{0x0c, AC_VERB_SET_AMP_GAIN_MUTE, AMP_IN_UNMUTE(1)},
@@ -5407,6 +5426,7 @@ static const char *alc882_models[ALC882_
	[ALC882_ARIMA]		= "arima",
	[ALC882_W2JC]		= "w2jc",
	[ALC885_MACPRO]		= "macpro",
+	[ALC885_APPLETV]	= "appletv",
	[ALC885_MBP3]		= "mbp3",
	[ALC882_AUTO]		= "auto",
};
@@ -5493,6 +5513,17 @@ static struct alc_config_preset alc882_p
		.channel_mode = alc882_ch_modes,
		.input_mux = &alc882_capture_source,
	},
+	[ALC885_APPLETV] = {
+		.mixers = { alc885_appletv_mixer },
+		.init_verbs = { alc885_appletv_init_verbs },
+		.num_dacs = ARRAY_SIZE(alc882_dac_nids),
+		.dac_nids = alc882_dac_nids,
+		.dig_out_nid = ALC882_DIGOUT_NID,
+		.dig_in_nid = ALC882_DIGIN_NID,
+		.num_channel_mode = ARRAY_SIZE(alc882_ch_modes),
+		.channel_mode = alc882_ch_modes,
+		.input_mux = &alc882_capture_source,
+	},
	[ALC882_TARGA] = {
		.mixers = { alc882_targa_mixer, alc882_chmode_mixer,
				alc882_capture_mixer },
@@ -5642,6 +5673,10 @@ static int patch_alc882(struct hda_codec
		case 0x106b0c00: /* Mac Pro */
			board_config = ALC885_MACPRO;
			break;
+		case 0x106b2800: /* AppleTV */
+			printk(KERN_INFO "hda_codec: Detected AppleTV RealTek Subsystem ID\n");
+			board_config = ALC885_APPLETV;
+			break;
		case 0x106b2c00: /* Macbook Pro rev3 */
			board_config = ALC885_MBP3;
			break;
@@ -5674,6 +5709,11 @@ static int patch_alc882(struct hda_codec
		alc882_gpio_mute(codec, 1, 0);
	}

+		if (board_config == ALC885_APPLETV) {  
+			alc882_gpio_mute(codec, 0, 0);
+			alc882_gpio_mute(codec, 1, 0);
+		}
+
	spec->stream_name_analog = "ALC882 Analog";
	spec->stream_analog_playback = &alc882_pcm_analog_playback;
	spec->stream_analog_capture = &alc882_pcm_analog_capture;